framework/src/meshgenerators/XYZDelaunayGenerator.C

```diff
@@ -340,36 +420,95 @@ XYZDelaunayGenerator::generate()
+    libMesh::MeshSerializer serial_hole(hole_mesh);
 
-            // I'd love to assert that we don't have any missing
-            // matches, but our holes might themselves have holes
-            if (it != mesh_faces.end())
+    // We'll look for any sides that match between the hole mesh and
+    // the newly triangulated mesh, and apply bcids accordingly.
+    for (auto elem : hole_mesh.element_ptr_range())
+      for (auto s : make_range(elem->n_sides()))
+        if (!elem->neighbor_ptr(s))
+        {
+          auto points = sorted_point_tuple(*elem, s);
+          auto it = mesh_faces.find(points);
+
+          // I'd love to assert that we don't have any missing
+          // matches, but our holes might themselves have holes
+          if (it != mesh_faces.end())
+          {
+            auto [main_elem, main_side] = it->second.first;
+            if (hole_i < _stitch_holes.size() && _stitch_holes[hole_i])
             {
-              auto [main_elem, main_side] = it->second;
               hole_boundary_info.add_side(elem, s, new_hole_bcid);
               mesh_boundary_info.add_side(main_elem, main_side, inner_bcid);
             }
+            // We would like to take this opportunity to identify the hole boundary
+            // We set the boundary id to hole_i + 1 at this stage (the default)
+            mesh_boundary_info.add_side(main_elem, main_side, hole_i + 1);
+            it->second.second = true;
           }
+        }
+  }
 
+  // Any sides that do not match the hole surfaces belong to the external boundary
+  // We set the boundary id to 0 at this stage (the default)
+  for (auto & [points, elem_side] : mesh_faces)
+    if (!elem_side.second)
+    {
+      auto [main_elem, main_side] = elem_side.first;
+      mesh_boundary_info.add_side(main_elem, main_side, 0);
+    }
+
+  // Now hole boundary ids are 1,2,..., N
+  // Now external boundary id is 0
+  // We will find the upper bound of the boundary ids of the mesh so that we would not overwrite
+  // things during renumbering (because of the bcids used for stitching, this is different from
+  // end_bcid)
+  const auto temp_bcid_shift = MooseMeshUtils::getNextFreeBoundaryID(*mesh);
+  // If we stitch holes, we want to make sure the netgen mesh has no conflicting boundary ids
+  // with the hole meshes.
+  // If we assign custom boundary ids, we want to make sure no overwriting occurs.
+  // So we shift all these boundary ids by temp_bcid_shift + free_boundary_id
+  // before we assign the new boundary ids
+  if (_stitch_holes.size() || output_boundary_id.size())
+    libMesh::MeshTools::Modification::change_boundary_id(
+        *mesh, 0, temp_bcid_shift + free_boundary_id);
+  if (_stitch_holes.size() || hole_boundary_ids.size())
+    for (auto hole_i : index_range(_hole_ptrs))
+      libMesh::MeshTools::Modification::change_boundary_id(
+          *mesh, hole_i + 1, hole_i + 1 + temp_bcid_shift + free_boundary_id);
+
+  // Now we can reassign the boundary ids
```

Comment:
why do we have to do this in two steps? It seems like could just do it in one step, going from 0 to  output_boundary_id[0 ?